sstables: scylla_metadata: add sstable identifier

bhalevy opened the pull request:
Keep a copy of the sstable uuid generation in a new
scylla_metadata sstable_identifier attribute.

If the SSTable happens to have a numerical generation
just create a new time-uuid and log a message about that.

Dump this new attribute in scylla sstable dump tool.

And add a unit test to verify that the written (and then
loaded) sstable identifier matches the sstable's generation.

The motivatrion for this change stems from backup
deduplication.  In essence, an sstable may already have been
backed up in a previous snapshot, and we don't want to
abck it up again if it's already present on external storage.

Today this is based on rclone that compares files checksums,
but once scylla will backup the sstables using the native
object-storage stack (#19890), we would like to use the sstable
globally-unique identifier for deduplication.  Although the
uuid-generation is encoded in the sstable path, the latter
may change, e.g. due to intra-node migration, so keep a copy
the original unique identifier in scylla-metadata, and that
attribute would survive file-based or intra-node migrations.

Fixes scylladb/scylladb#20459

* Optional enhancement.  no backport required

bhalevy commented:
> > Keep a copy of the sstable uuid generation in the scylla_metadata component.
> > If the SSTable happens to have a numerical generation just create a new time-uuid generation and log a warning.
> > Dump this new attribute in scylla sstable dump tool.
> > And add a unit test to verify that the written (and then loaded) original sstable generation matches the sstable's generation.
> > Fixes #20459
> > 
> > * Optional enhancement.  no backport required
> 
> I remember reading about the motivation somewhere, but I forgot about it. I think it would good to add it here to the cover letter.

okay, makes sense.  The gist of it is for backup deduplication, so we can identify sstables that changed their generation, e.g. via intra-node migration, but they are equivalent, i.e. they contain the same data.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12254/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12254/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_datafile_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12254/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1049/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4108/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12125/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12254/)
#### Build Details:
- Duration: 9 hr 4 min
- Builder: spider8.cloudius-systems.com

denesb commented on sstables/sstables.cc:
```diff
@@ -1391,6 +1391,12 @@ future<> sstable::update_info_for_opened_data(sstable_open_config cfg) {
+    }
+    if (cfg.load_first_and_last_position_metadata) {
+        co_await load_first_and_last_position_in_partition();
+    }
```
Why is this duplicated here? I already see this snippet above.

bhalevy replied:
weird. maybe cut-and-paste pilot error.
will remove the redundant code.

denesb commented on sstables/sstables.cc:
```diff
@@ -1872,6 +1878,15 @@ sstable::write_scylla_metadata(shard_id shard, sstable_enabled_features features
+    if (generation().is_uuid_based()) {
+        gen = generation().as_uuid();
+    } else {
+        sstlog.warn("SSTable generation {} expected to be UUID based when writing scylla metadata", generation());
```
Why the warning? Is it an error to have non-uuid generation for an sstable?

bhalevy replied:
We are generating a unique generation here, so it may come as a surprise.
Maybe `info` level would be more appropriate.

That said, we should probably deprecate the `uuid_sstable_identifiers_enabled` option and stop supporting it as soon as possible.

denesb replied:
I think as long as numeric generations are valid, we should not log anything here. Why would it be a surprise that we have a numeric generation, which is still a perfectly valid option?

denesb commented on sstables/sstables.hh:
```diff
@@ -968,6 +969,10 @@ public:
         return _origin;
     }
 
+    const generation_type& original_sstable_generation() const noexcept {
```
Shouldn't this be optional or something? Not all sstables will have this.

bhalevy replied:
The default constructed generation_type evaluates to `false`, so it will be engaged only if it was loaded from `scylla_metadata`

xemul replied:
This sounds like good opportunity for std::optimized_optional<> , it will make it immediately obvious that this generation can validly be missing

denesb replied:
Agree, in general we encourage using `optional<>` even when the type has a distinct "empty" state. Using `optional<>` wrapper makes the intended interface clear.

bhalevy replied:
Done in v4 ([633ca64](https://github.com/scylladb/scylladb/commit/633ca645486fa7314e5a1dd73c31141524cb4552))

denesb commented on sstables/types.hh:
```diff
@@ -593,7 +604,8 @@ struct scylla_metadata {
             disk_tagged_union_member<scylla_metadata_type, scylla_metadata_type::ScyllaVersion, scylla_version>,
-            disk_tagged_union_member<scylla_metadata_type, scylla_metadata_type::ExtTimestampStats, ext_timestamp_stats>
+            disk_tagged_union_member<scylla_metadata_type, scylla_metadata_type::ExtTimestampStats, ext_timestamp_stats>,
+            disk_tagged_union_member<scylla_metadata_type, scylla_metadata_type::SSTableGeneration, sstable_generation>
```
I think this is confusing name, the sstable already has a generation, why is there a second one in the metadata?
I think sstable identifier would be a more appropriate name, that is what we are using this UUID for.

bhalevy replied:
makes sense, will change, thanks.

bhalevy commented:
In v2 ([b11e1c8](https://github.com/scylladb/scylladb/commit/b11e1c883d98d3094be776e2b74aac3c790e36bf)):
- update cover letter to include motivation
- renamed scylla_metadata member to sstable_identifier
- defined sstable_id using tagged_uuid strong type
  - added comments about it being null only when unset
- change log warning to info level, printing the newly created identifier, when the sstable has a numerical generation.

denesb commented on sstables/sstables.cc:
```diff
@@ -1872,6 +1875,15 @@ sstable::write_scylla_metadata(shard_id shard, sstable_enabled_features features
+        sid = sstable_id(generation().as_uuid());
+    } else {
+        sid = sstable_id(utils::UUID_gen::get_time_UUID());
+        sstlog.info("SSTable generation {} is numerical. SSTable identifier in scylla_metadata set to {}", generation(), sid);
```
To be useful, this log message should contain enough information about the sstable, so that we can identify it. Best to include its path in the message.

bhalevy replied:
Done in [1f1c797](https://github.com/scylladb/scylladb/commit/1f1c7972194188d8f5cf96ee249bb5ba4d1da9b2)

bhalevy commented:
In v4 ([633ca64](https://github.com/scylladb/scylladb/commit/633ca645486fa7314e5a1dd73c31141524cb4552)):
- used `optimized_optional<sstable_id>` for the _sstable_identifer member
  - adjusted callers respectively

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12275/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12275/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_datafile_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12275/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1056/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12133/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4116/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12275/)
#### Build Details:
- Duration: 7 hr 43 min
- Builder: i-08b5b2afada96e8bb (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12300/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12300/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_datafile_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12300/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1071/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4133/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12152/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12300/)
#### Failed Tests (1/36796):
- [test_tablet_drain_failure_during_decommission.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12300/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_dev_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.4)
#### Build Details:
- Duration: 4 hr 6 min
- Builder: spider8.cloudius-systems.com
